### PR TITLE
QueryState::new_with_state

### PR DESCRIPTION
**Pull request by Suficio**
# Objective

Allows providing state to `QueryState` which requires input other than `WorldQuery::init_state(&mut World)` to be constructed.

This allows the user to provide a custom state such as `ComponentId` for dynamic queries which cannot be derived from the type definition.

## Solution

Create new method `QueryState::new_with_state`, which is now called by `QueryState::new` using `WorldQuery::init_state`.

### CODE REVIEW

File: crates/bevy_ecs/src/query/state.rs [hunk 1]
```diff
@@ -96,6 +96,15 @@ impl<Q: WorldQuery, F: ReadOnlyWorldQuery> QueryState<Q, F> {
         let fetch_state = Q::init_state(world);
         let filter_state = F::init_state(world);
 
+        Self::new_with_state(world, fetch_state, filter_state)
+    }
+
+    /// Creates a new [`QueryState`] with custom initial state
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/query/state.rs:102]: This could use a doc test; it's not going to be clear to users how or why to use this method.

Comment (Suficio) [reply crates/bevy_ecs/src/query/state.rs]: Done

**Review by alice-i-cecile — CHANGES_REQUESTED** [review]
The example is great, but we probably also want a third query_state initialization showing how to init filter state.

**Review by alice-i-cecile — APPROVED** [review]

**Comment by alice-i-cecile** [comment]
Awesome work; thanks for the prompt response. This is a niche feature, but I think it will be useful as part of our "dynamic components" story.

**Comment by Suficio** [comment]
> Awesome work; thanks for the prompt response. This is a niche feature, but I think it will be useful as part of our "dynamic components" story.

No problem, your prompt response is much more appreciated :)

With this feature it is already possible to make a dynamic query by implementing `WorldQuery` on a `ComponentId`.

**Review by zicklag — COMMENTED** [review]
I'm not qualified enough to review the Bevy query internals yet/anymore, but I seem to remember needing this in my now very old scripting API/PR, so.., 👍 😁

**Comment by alice-i-cecile** [comment]
Yep, I mostly wanted to hear if you thought it was useful :) The implementation is trivial, and the docs are great.

**Comment by Suficio** [comment]
Nice catch @BoxyUwU!

While there is nothing unsafe that this function does, indeed we can incorrectly cast types as a consequence. There are two ways I see that this can be resolved.

1. Validate `ComponentId` for `&T` queries during `<&T as WorldQuery>::init_fetch`
2. Make `QueryState::new_with_state` unsafe

**Comment by alice-i-cecile** [comment]
The former is definitely safer. This would be a one time cost paid during system / query initialization, correct?

**Comment by Suficio** [comment]
Made the method unsafe and added the appropriate safety comment

**Comment by Suficio** [comment]
> Currently the # Safety comment only shows a bad example and suggest using safer methods instead, usually these comments also contain an exhautive list of things you need to validate so that the call is guaranteed safe.

It's sort of difficult because this method is only unsafe when querying component references, but I reordered the safety comment to more strongly word the requirements for safety.

> As I understand it, the point of the new_with_state method is to be able to be used if you only know a few ComponentIds at runtime and want to get Ptr<'_>s to them, instead of specifying T to get &Ts, right?

This is exactly right! However, this isn't the only use, one could likely implement world query for `Vec<T>`.

Your solution is definitely and improvement, however, I would like to prioritize minimal changes with this PR.

Dynamic query work is still up for discussion on how it should be implemented, and I would like to avoid establishing any future design directions, and instead, work with the API surface as it is available today.

**Comment by BoxyUwU** [comment]
> It's sort of difficult because this method is only unsafe when querying component references

I disagree with this, worldquery impls should be able to rely on the state being created from their own `init_state` fn and nothing else. More than just "unsafe" it should be unsound to call `new_with_state` with a value that isnt from `init_state` unless the worldquery explicitly relaxes this in its documentation.

If the above is true then this API is also kind of Not Great, I would much prefer the solution @jakobhellermann proposes with a `type Config;` since we could probably make that API safe

**Comment by Suficio** [comment]
The problem I see is that `WorldQuery` tuple implementations will require all elements to have `Config = ()`, due to system params requiring `Config = ()` (as systems are not able to provide the config), making queries with arbitrary `Config` unusable for the majority of use patterns.

I've tried the implementation suggested and it results in **a lot** of extra type constraints that need to be managed across the codebase, and due to the above, severely restricts the reusability of existing implementations.

> More than just "unsafe" it should be unsound to call new_with_state with a value that isnt from init_state unless the worldquery explicitly relaxes this in its documentation.

Due to this restriction on `Config = ()`, I've slightly relaxed the safety comment. 

While I agree that queries should be able to rely on their `init_state`, it is not possible to initialize state using only `&World`, and the suggested `Config` approach is not able to facilitate the use cases I need.

**Comment by Suficio** [comment]
I think one could argue that having a default here is equally dangerous as providing the state outright, due to the potential implications of accidentally using a `ComponentId` default for `Ptr`s.

It results in a situation where you don't know whether `init_state` config has correct data or whether it needs to be inferred otherwise. I guess we could use an `Option` but this is still added type complexity for a very niche benefit that this PR brings.

The added benefit of this PR would be that you can avoid the added complication of manipulating type bounds.

**Comment by Suficio** [comment]
Resolved by #9774

**Comment by Suficio** [comment]
Resolved by #9774

**Pull request closed without merging**